### PR TITLE
unix socket.settimeout()

### LINKED ISSUE #2379: unix port socket.settimeout not implemented
`socket.settimeout()` isn't implemented in the UNIX port, which makes it extremely difficult to test code using socket timeouts on UNIX.

(This is especially fun because the ESP8266 doesn't have `uselect`, I have no platform independent way to regain control.)

### PR DESCRIPTION
**Pull request by danni**
Implements `socket.settimeout()` for unix.

Fixes #2379

### CODE REVIEW

**Comment by dbrignoli** [comment]
Hi @danni, I'll let others comment on if raising an exception on timeout is the desired behaviour here. Assuming raising an exception is desirable I would think socket_read() and socket_write() should change _errcode_ to _MP_ETIMEDOUT_ only if the timeout was set. (my 2c)

**Comment by pfalcon** [comment]
Please see comment https://github.com/micropython/micropython/issues/2379#issuecomment-244386899

File: unix/modsocket.c [hunk 1]
```diff
@@ -99,6 +99,11 @@ STATIC mp_uint_t socket_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *errc
     mp_int_t r = read(o->fd, buf, size);
     if (r == -1) {
         *errcode = errno;
+
+        if (*errcode == EAGAIN) {
```
Comment (pfalcon) [inline unix/modsocket.c]: This errno conversion is not correct for a case of non-blocking socket (for it, EAGAIN should remain EAGAIN).

**Comment by pfalcon** [comment]
Based on all the follow-up experience, I now think that it makes sense to add settimeout() to Unix port, after all. The reason is that without it, it's hard (up to impossibility without kludges) to write reliable *synchronous* applications. The original suggestion (https://github.com/micropython/micropython/issues/2379) was to use uselect.poll() instead of settimeout(), but in synchronous stream architecture, sometimes there's nowhere to put that uselect.poll(). For example, we can take socket-based stream and pass it to ujson.load(). That's the reason for me giving up on this long-standing question.

But note that for truly async apps, .settimeout is useless, so I'd suggest to implement a config option to disable it (and save code bytes).

Now about specific patch: it uses POSIX-mandated SO_RCVTIMEO, SO_SNDTIMEO setsockopt: http://pubs.opengroup.org/onlinepubs/7908799/xns/setsockopt.html , and thus avoids adding more state on uPy side (i.e. new fields to mp_obj_socket_t). That's good news. Bad news is that this patch is not entirely correct, it breaks non-blocking sockets. Given that realistic minimum for alloc unit in uPy is 16 bytes, it actually makes sense to cache blocking/non-blocking status in mp_obj_socket_t (which will grow from 8 to 12 bytes, still in 1 alloc block), and use that to translate EAGAIN into MP_ETIMEDOUT or not.

Another issue this patch has is lack of correlation between settimeout(0) and setblocking(False). I'd consider that minor, and acceptable measure to save code bytes again. But even better solution would be: a) make settimeout() support dependent on option in mpconfig.h; b) if enabled, treat settimeout(0) as setblocking(False).

**Comment by pfalcon** [comment]
> it makes sense to add settimeout() to Unix port, after all.

Updated solution, with this patch, and fixes on top of it, is posted as https://github.com/micropython/micropython/pull/4219

**Comment by dpgeorge** [comment]
This was merged in 80a25810f9b246094b31ce7d465a0f25842c9be6.  Thanks @danni for the initial patch.

**Pull request closed without merging**